API/Controllers/TournamentsController.cs

```diff
@@ -20,68 +22,89 @@ IAdminNoteService adminNoteService
-    /// <returns>
-    /// A page of tournaments
-    /// </returns>
     [HttpGet]
     [ProducesResponseType<IEnumerable<TournamentDTO>>(StatusCodes.Status200OK)]
-    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
-    public async Task<IActionResult> ListAsync([FromQuery] TournamentRequestQueryDTO? requestQuery)
-    {
-        requestQuery ??= new TournamentRequestQueryDTO();
-
-        ICollection<TournamentDTO> tournaments = await tournamentsService.GetAsync(requestQuery);
-        return Ok(tournaments);
-    }
+    public async Task<IActionResult> ListAsync([FromQuery] TournamentRequestQueryDTO requestQuery) =>
+        Ok(await tournamentsService.GetAsync(requestQuery));
 
     /// <summary>
     /// Submit a tournament
     /// </summary>
     /// <param name="tournamentSubmission">Tournament submission data</param>
     /// <response code="400">
-    /// If the given <see cref="tournamentSubmission"/> is malformed or
-    /// if a tournament matching the given name and ruleset already exists
+    /// The tournament submission is malformed or
+    /// a tournament matching the given name and ruleset already exists
     /// </response>
     /// <response code="201">Returns location information for the created tournament</response>
     [HttpPost]
     [Authorize(Roles = OtrClaims.Roles.User)]
     [ProducesResponseType(StatusCodes.Status400BadRequest)]
     [ProducesResponseType<TournamentCreatedResultDTO>(StatusCodes.Status201Created)]
-    public async Task<IActionResult> CreateAsync([FromBody] TournamentSubmissionDTO tournamentSubmission)
+    public async Task<IActionResult> CreateAsync([FromBody][Required] TournamentSubmissionDTO tournamentSubmission)
     {
-        if (!ModelState.IsValid)
-        {
-            return BadRequest(ModelState.ErrorMessage());
-        }
+        // Sanitize id collections
+        tournamentSubmission.Ids = tournamentSubmission.Ids.Distinct().Where(id => id >= 1);
+        tournamentSubmission.BeatmapIds = tournamentSubmission.BeatmapIds.Distinct().Where(id => id >= 1);
 
+        #region Model Validation
+
+        // Check for unique submissions
         if (await tournamentsService.ExistsAsync(tournamentSubmission.Name, tournamentSubmission.Ruleset))
         {
-            return BadRequest(
-                $"A tournament with name {tournamentSubmission.Name} for ruleset {tournamentSubmission.Ruleset} already exists"
+            ModelState.AddModelError(
+                nameof(TournamentSubmissionDTO.Name),
+                $"A tournament with the name '{tournamentSubmission.Name}' and ruleset " +
+                $"'{tournamentSubmission.Ruleset}' already exists."
+            );
+            ModelState.AddModelError(
+                nameof(TournamentSubmissionDTO.Ruleset),
+                $"A tournament with the ruleset '{tournamentSubmission.Ruleset}' and name " +
+                $"'{tournamentSubmission.Name}' already exists."
```

Comment:
I think it would be good to explain in a comment why it's necessary to add two separate `ModelState` errors because it looks redundant at a glance.

Reply:
The `ModelState` property is of the type `ModelStateDictionary` which, as it implies, is literally just a dictionary of keys to values. Keys being the names of model properties and values being any number of error messages. Since this case shows that both items in the compound unique index are invalid, I think they should both receive a model state error. Ill add a short comment too